File: pr1-java/src/main/java/services/Service.java
package services;

import domain.Game;
import domain.Ticket;
import domain.User;
import domain.exceptions.DuplicateException;
import domain.exceptions.LogInException;
import domain.exceptions.NotFoundException;
import domain.exceptions.ParameterException;
import repository.GameRepository;
import repository.TicketRepository;
import repository.UserRepository;

import java.util.ArrayList;
import java.util.Collection;

public class Service {
    private final UserRepository userRepository;
    private final GameRepository gameRepository;
    private final TicketRepository ticketRepository;

    public Service(UserRepository userRepository, GameRepository gameRepository, TicketRepository ticketRepository) {
        this.userRepository = userRepository;
        this.gameRepository = gameRepository;
        this.ticketRepository = ticketRepository;
    }

    /**
     * Logs in a user
     *
     * @param username the username of the user to be logged in
     * @return the logged in user
     * @throws NotFoundException if user not found
     * @throws LogInException    if user already logged in
     */
    public User logInUser(String username) throws NotFoundException, LogInException {
        User user = userRepository.getOne(username);
        if (user.getStatus().equals("logged-in"))
            throw new LogInException("user already logged in");
        userRepository.setUserStatus(user.getId(), "logged-in");
        return user;
    }

    /**
     * Logs out a user
     *
     * @param username the username of the user to be logged out
     * @return the logged out user
     * @throws NotFoundException if user not found
     * @throws LogInException    if user already logged out
     */
    public User logOutUser(String username) throws NotFoundException, LogInException {
        User user = userRepository.getOne(username);
        if (user.getStatus().equals("logged-out"))
            throw new LogInException("user already logged out");
        userRepository.setUserStatus(user.getId(), "logged-out");
        return user;
    }

    /**
     * Signs up a user
     *
     * @param username the username of the user to be added
     * @return the signed up user
     * @throws DuplicateException if the user already exists
     */
    public User signUpUser(String username) throws DuplicateException {
        User user = new User(username, "logged-in");
        user = userRepository.add(user);
        return user;
    }

    public Collection<Game> getAllGames() {
        Collection<Game> games = new ArrayList<>();
        for (Game game : gameRepository.getAll())
            games.add(game);
        return games;
    }

    public void sellSeats(Game game, String clientName, Integer seatsCount) throws ParameterException, NotFoundException {
        if (game.getAvailableSeats() < seatsCount)
            throw new ParameterException("not enough seats available");
        gameRepository.setGameAvailableSeats(game.getId(), game.getAvailableSeats() - seatsCount);
        while (seatsCount != 0) {
            try {
                ticketRepository.add(new Ticket(game, clientName));
            } catch (DuplicateException ignored) {
            }
            seatsCount--;
        }
    }

    public Collection<Game> getGamesWithAvailableSeatsDescending() {
        Collection<Game> games = new ArrayList<>();
        for (Game game : gameRepository.getGamesByAvailableSeatsDescending(true))
            if (game.getAvailableSeats() > 0)
                games.add(game);
        return games;
    }
}

File: pr1-cs/Services/Service.cs
using System.Collections.Generic;
using pr1_cs.Domain;
using pr1_cs.Domain.Exceptions;
using pr1_cs.Repository;

namespace pr1_cs.Services
{
    public class Service
    {
        private readonly IGameRepository _gameRepository;
        private readonly ITicketRepository _ticketRepository;
        private readonly IUserRepository _userRepository;

        public Service(IUserRepository userRepository, IGameRepository gameRepository,
            ITicketRepository ticketRepository)
        {
            _userRepository = userRepository;
            _gameRepository = gameRepository;
            _ticketRepository = ticketRepository;
        }

        public User LogInUser(string username)
        {
            var user = _userRepository.GetOne(username);
            if (user.Status == "logged-in")
                throw new LogInException("user already logged in");
            _userRepository.SetUserStatus(user.Id, "logged-in");
            return user;
        }

        public User LogOutUser(string username)
        {
            var user = _userRepository.GetOne(username);
            if (user.Status == "logged-out")
                throw new LogInException("user already logged out");
            _userRepository.SetUserStatus(user.Id, "logged-out");
            return user;
        }

        public User SignUpUser(string username)
        {
            var user = new User(username, "logged-in");
            user = _userRepository.Add(user);
            return user;
        }

        public IEnumerable<Game> GetAllGames()
        {
            return _gameRepository.GetAll();
        }

        public void SellSeats(Game game, string clientName, int seatsCount)
        {
            if (game.AvailableSeats < seatsCount)
                throw new ParameterException("not enough seats available");
            _gameRepository.SetGameAvailableSeats(game.Id, game.AvailableSeats - seatsCount);
            while (seatsCount != 0)
            {
                try
                {
                    _ticketRepository.Add(new Ticket(game, clientName));
                }
                catch (DuplicateException)
                {
                }

                seatsCount--;
            }
        }

        public IEnumerable<Game> GetGamesWithAvailableSeatsDescending()
        {
            var games = new List<Game>();
            foreach (var game in _gameRepository.GetGamesByAvailableSeatsDescending(true))
                if (game.AvailableSeats > 0)
                    games.Add(game);
            return games;
        }
    }
}